File: leetcode_py/Binary Tree Maximum Path Sum.py
# Definition for a  binary tree node
class TreeNode:
     def __init__(self, x):
         self.val = x
         self.left = None
         self.right = None


class Solution:
    def __init__(self):
        self.max = -1<<64

    def helper(self, root):
        if not root:
            return 0
        left = max(self.helper(root.left), 0)
        right = max(self.helper(root.right), 0)
        self.max = max(self.max, left + right + root.val)
        return max(left, right) + root.val

    def maxPathSum(self,root):
        self.helper(root)
        return self.max


class Solution:
    def __init__(self):
        self.max = -1<<64

    def helper(self, root):
        left = max(self.helper(root.left), 0)
        right = max(self.helper(root.right), 0)
        self.max = max(self.max, left + right + root.val)
        return max(left, right) + root.val

    def maxsum(self,root):
        self.helper(root)
        return self.max


class Solution:
    def f(self,root):
        if not root:
            return -1<<64,0
        lmax,l=self.f(root.left)
        rmax,r=self.f(root.right)
        tmp=max(l+root.val,r+root.val,root.val)
        return max(lmax,rmax,tmp,l+r+root.val),tmp

    # @param {TreeNode} root
    # @return {integer}
    def maxPathSum(self, root):
        return self.f(root)[0]


class Solution:
    def maxsum(self,root):
        if root is None:
            return -2<<64
        left=self.maxsum(root.left)
        right=self.maxsum(root.right)
        self.maxs=max(self.maxs,left,right,root.val+left+right)
        return max(root.val,root.val+left,root.val+right)
  
    # @param root, a tree node
    # @return an integer
    def maxPathSum(self, root):
        if root is None:
            return 0
        self.maxs=-2<<64
        return max(self.maxsum(root),self.maxs)

if __name__=="__main__":
    a=Solution()
    tree=TreeNode(-2)
    tree.left=TreeNode(1)
    print a.maxPathSum(tree)
    print a.maxPathSum(tree.left)

File: leetcode_js/Binary Tree Maximum Path Sum.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
// return
var helper = function(root, res){
    if(!root) return 0;
    var left = Math.max(0, helper(root.left, res));
    var right = Math.max(0, helper(root.right, res));
    res[0] = Math.max(res[0], root.val + left + right);
    return root.val + Math.max(left, right);
}
var maxPathSum = function(root) {
    var res = [root.val];
    helper(root, res);
    return res[0];
};


/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
// return
var helper = function(root, res){
    if(root.left && root.right){
        var leftmaxs = helper(root.left, res);
        var rightmaxs = helper(root.right, res);
        var maxsons = Math.max(leftmaxs, rightmaxs);
        var tmp = Math.max(root.val + leftmaxs + rightmaxs, root.val + Math.max(0, maxsons));
        res[0] = Math.max(res[0], tmp);
        return root.val + (maxsons>0?maxsons:0);
    }
    if(root.right){
        var rightmaxs = helper(root.right, res);
        var tmp = Math.max(root.val, root.val + rightmaxs);
        res[0] = Math.max(res[0], tmp);
        return root.val + (rightmaxs>0?rightmaxs:0);
    }
    if(root.left){
        var rightmaxs = helper(root.left, res);
        var tmp = Math.max(root.val, root.val + rightmaxs);
        res[0] = Math.max(res[0], tmp);
        return root.val + (rightmaxs>0?rightmaxs:0);
    }
    res[0] = Math.max(res[0], root.val);
    return root.val;
}
var maxPathSum = function(root) {
    if(!root) return 0;
    var res = [root.val];
    helper(root, res);
    return res[0];
};


var maxPathSum = function(root) {
    return helper(root)[0];
 };
function helper(root) {
    console.log(root.val + '  root')
    if(root.left == null && root.right == null) {
        return [root.val, root.val];
    }
    var maxall3 = - Number.MAX_VALUE;
    var maxsingle = - Number.MAX_VALUE;
    var maxall1, maxall2, maxleft, maxright;
    if(root.left != null) {
        [maxall1, maxleft] = helper(root.left);
        maxsingle = Math.max(maxsingle, root.val + maxleft, root.val);
        maxall3 = Math.max(maxall3, maxall1, maxsingle); 
    }
    if(root.right != null) {
        [maxall2, maxright] = helper(root.right);
        maxsingle = Math.max(maxsingle, root.val + maxright, root.val);
        maxall3 = Math.max(maxall3, maxall2, maxsingle); 
    }
   
    if(root.left != null && root.right != null) {
        maxsingle = Math.max(maxsingle, root.val + Math.max(maxleft, maxright), root.val);
        maxall3 = Math.max(maxall3, root.val + maxleft + maxright, maxsingle);
        
    }
   
    return [maxall3, maxsingle];
 }